HSEARCH-2071 Support EmbeddedId

DavideD opened the pull request:
https://hibernate.atlassian.net/projects/HSEARCH/issues/HSEARCH-2071

I checked the test `EmbeddableIdTest` and `ProgrammaticEmbeddedItTest` and they won't work with elasticsearch for the following reasons:
* the field bridge used for the id is not a `MetadataProvidingFieldBridge`; because of this, the new fields are not defined during the schema creation and this causes some errors when we need to create them;
* the fields name contain a `.` ;

I've created new tests without this limitations and checked that the Lucene and Elasticsearch backend are consistent. This resulted in another test failure because on Lucene a field was store not analyzed while in Elasticsearch was stored analyzed.

I tried to fix the issue by adding some information in the `BridgeDefinedField`. Let me know if it makes sense.

Sanne commented:
@DavideD why is this assigned to you? Please give yourself advice in how to proceed :)

DavideD commented:
Problem solved :)

Sanne commented:
ouch :) :+1:

gsmet commented:
Jenkins, please retest this

gsmet commented:
Jenkins, retest this please

gsmet commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchIndexManager.java:
```diff
@@ -383,7 +390,11 @@ private void addFieldMapping(JsonObject payload, BridgeDefinedField bridgeDefine
+			String index = getIndex( bridgeDefinedField );
+			field.addProperty( "index", index );
+
+			// TODO: We need to add the analyzer name
```
Looks like this TODO should be fixed.

gsmet commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchIndexManager.java:
```diff
@@ -439,7 +450,22 @@ private String getIndex(EntityIndexBinding binding, DocumentFieldMetadata fieldM
+	@SuppressWarnings("deprecation")
+	private String getIndex(BridgeDefinedField bridgeDefinedField) {
+		// Never analyze boolean
+		if ( bridgeDefinedField.getType().equals( FieldType.BOOLEAN ) ) {
```
Same for dates and numerics, see a few lines above.

gsmet commented on engine/src/main/java/org/hibernate/search/engine/metadata/impl/AnnotationMetadataProvider.java:
```diff
@@ -1271,6 +1255,22 @@ private void bindFieldAnnotation(
+				propertyMetadataBuilder.addBridgeDefinedField( field );
+			}
+		}
+
```
Any particular reason why you moved this code around?

DavideD replied:
I remember there was one, but I forgot now :)

I will double check

DavideD replied:
Now getBridgeContributedFieldMetadata uses the DocumentFieldMetada to make sure that the default used for the custom fields are the same as the one declare on the main field.
For this reason I moved the code after the creation of the metadata.

gsmet commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java:
```diff
@@ -77,4 +79,7 @@
 
+	@Message(id = ES_BACKEND_MESSAGES_START_ID + 11,
+			value = "Projection of non-JSON-primitive field values is not supported: '%1$s'")
+	SearchException unsupportedProjectionOfNonJSONPrimitiveFields(JsonElement value);
```
Minor nitpicking: we use the case "Json" everywhere as we conform to the Gson convention. I'm only talking about the method name, the message is fine.

DavideD replied:
ok

gsmet commented on orm/src/test/java/org/hibernate/search/test/id/withmeta/EmbeddedIdWithMetadataProvidingBridgeTest.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	/**
+	 * HSEARCH-HSEARCH-306, HSEARCH-248
```
It's probably coming from the original test but s/HSEARCH-HSEARCH/HSEARCH/.

DavideD replied:
ok

gsmet commented:
Couple of typos in one of the commit message:
> HSEARCH-2071 Use correct defaults for fields provided by field bridges
>  Custom field bridges in the Lucene backend can create new fields using as default
>  the settings for property accessor,

s/,/./

>  For example, if the fieldbridge is called for an id, the custom fields won't be analyzed; when
>  used an a normal field they are going to be analyzed.

s/fieldbridge/field bridge/
s/used an a/used as a/

gsmet commented on orm/src/test/java/org/hibernate/search/test/id/withmeta/PersonPKMetadataProviderBridge.java:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	private static String firstName(String name) {
+		return name + "_firstName";
```
OK, so in fact, you can have fields with dots: we treat them as nested documents which sounds logical. You probably should remove this part from the commit message.

DavideD replied:
I will rephrase the sentence, what I mean is that if you create a new field in the field bridge the name does not have to contain a `.`.

gsmet commented:
Finished with the review. Feel free to ping me if there's something unclear.

DavideD commented:
Fixed the remarks, added the analyzer name and rebased

gsmet commented:
@DavideD I'm not convinced it's a good idea to add the analyzer and the analyzed/not analyzed information to the BridgeDefinedField.

IIRC, in the Lucene case, the only bridge defined field inheriting from the analyzer definition is the one having the same name of the default name. The other fields are not analyzed.

If I'm correct on this, I think we should stick with it for now and have the exact same behavior in Elasticsearch.

That being said, I would really like to revisit this but in this case, we also need to add the ability to override the analyzer definition when the bridge defined field is defined.

So for now, I'll stick to the behavior we already have in Lucene and create a JIRA with the additional patch attached (or a link to one of your branches) so that we can build on it later. I had a couple of thoughts in this area, I have to dig a bit more a bug I found and I'll write something on the list.

By the way, looks like your new tests are not passing.

DavideD commented:
@gsmet I've force-pushed  some changes:

1.  I've removed the part related to the analyzer name
2. Added a new test for core and elasticsearch that make sure that the defaults applied are the same (there is duplication and it might be improved but I don't want to spend more time on it). The default analyzer is used as yuo says.

I think we need a new JIRA to allow to set an analyzer for custom field. WWDYT?

gsmet commented:
Merged! Thanks @DavideD!
